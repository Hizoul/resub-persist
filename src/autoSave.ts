import persist from "./persist"
import { IPersistableStore, KeyOrKeys } from "./type"

const alreadySaving: {[index: string]: boolean | undefined} = {}

const redoSearch: {[index: string]: boolean | undefined} = {}

const tryToSaveStore = (storage: any, store: IPersistableStore)  => {
  const doSave = async () => {
    if (alreadySaving[store.name]) {
      redoSearch[store.name] = true
      return
    }
    alreadySaving[store.name] = true
    await persist(storage, [store])
    alreadySaving[store.name] = false
    if (redoSearch[store.name]) {
      redoSearch[store.name] = false
      doSave()
    }
  }
  return doSave
}

/**
 * Subscribe to `store` with the optional `key` param. 
 * Calls `persist` for the store after every `resub` trigger relevant to `key`.
 * @param storage pass localforage or AsyncStorage here
 * @param store pass your IPersistableStore that should be persisted on change
 * @param key define to which `resub`-keys the autosave should react
 */
const autoSave = (storage: any, store: IPersistableStore, key?: KeyOrKeys) => {
  const s: any = store
  return s.subscribe(tryToSaveStore(storage, store), key)
}

export default autoSave
