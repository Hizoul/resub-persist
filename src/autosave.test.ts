import autoSave from "./autoSave"
import memForage from "./testUtil/miniMemForage"
import TestStore1 from "./testUtil/testStore"

const waitFor = (milli: number) => {
  return new Promise((resolve, reject) => {
    setTimeout(resolve, milli)
  })
}


test("Cover Persist Proeprty Test", async () => {
  expect(memForage.getData()).toMatchSnapshot("empty before any actions")
  TestStore1.toggleMyBool()
  TestStore1.setMyString("dont save me yet")
  await waitFor(50)
  expect(memForage.getData())
  .toMatchSnapshot("nothing automatically persisted")
  const sub = autoSave(memForage, TestStore1)
  TestStore1.toggleMyBool()
  TestStore1.setMyString("i was autosaved")
  await waitFor(50)
  const data = memForage.getData()
  expect(memForage.getData())
  .toMatchSnapshot("automatic persistence worked")
  TestStore1.unsubscribe(sub)
  TestStore1.toggleMyBool()
  TestStore1.setMyString("i shouldn't be saved")
  await waitFor(50)
  expect(memForage.getData())
  .toMatchSnapshot("disabling automatic saving results in no changes made")
  expect(JSON.stringify(data) === JSON.stringify(memForage.getData())).toBeTruthy()
  const sub2 = autoSave(memForage, TestStore1)
  TestStore1.toggleMyBool()
  TestStore1.setMyString("autochangesaved again")
  await waitFor(50)
  expect(memForage.getData())
  .toMatchSnapshot("resubbing autosaves again")
  TestStore1.toggleMyBool()
  await waitFor(200)
  expect(memForage.getData())
  .toMatchSnapshot("boolean can become true")
})
