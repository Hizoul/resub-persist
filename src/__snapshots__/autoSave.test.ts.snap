// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`automatic persistence worked 1`] = `
Object {
  "store.detailList": "{\\"myBool\\":false,\\"myString\\":\\"i was autosaved\\"}",
}
`;

exports[`boolean can become true 1`] = `
Object {
  "store.detailList": "{\\"myBool\\":true,\\"myString\\":\\"autochangesaved again\\"}",
}
`;

exports[`disabling automatic saving results in no changes made 1`] = `
Object {
  "store.detailList": "{\\"myBool\\":false,\\"myString\\":\\"i was autosaved\\"}",
}
`;

exports[`empty before any actions 1`] = `Object {}`;

exports[`nothing automatically persisted 1`] = `Object {}`;

exports[`resubbing autosaves again 1`] = `
Object {
  "store.detailList": "{\\"myBool\\":false,\\"myString\\":\\"autochangesaved again\\"}",
}
`;
